sklearn/metrics/tests/test_score_objects.py

```diff
@@ -1289,10 +1289,23 @@ def test_PassthroughScorer_metadata_request():
     # Test that _PassthroughScorer doesn't change estimator's routing.
     assert_request_equal(
         scorer.get_metadata_routing(),
-        {"fit": {"sample_weight": True}, "score": {"sample_weight": "alias"}},
+        {"score": {"sample_weight": "alias"}},
     )
 
 
+@pytest.mark.usefixtures("enable_slep006")
+def test_PassthroughScorer_set_score_request():
+    """Test that _PassthroughScorer.set_score_request adds the correct metadata request
+    on itself."""
+    meta_est = GridSearchCV(estimator=LinearSVC(), param_grid={"C": [0.1, 1]})
+
+    # make a `_PassthroughScorer` with `check_scoring`:
+    scorer = check_scoring(meta_est, None)
+    scorer.set_score_request(sample_weight=True)
+
+    assert str(scorer.get_metadata_routing()) == str(scorer._metadata_request)
```

Comment:
This could be made more minimal:

```patch
--- a/sklearn/metrics/tests/test_score_objects.py
+++ b/sklearn/metrics/tests/test_score_objects.py
@@ -1297,13 +1297,11 @@ def test_PassthroughScorer_metadata_request():
 def test_PassthroughScorer_set_score_request():
     """Test that _PassthroughScorer.set_score_request adds the correct metadata request
     on itself."""
-    meta_est = GridSearchCV(estimator=LinearSVC(), param_grid={"C": [0.1, 1]})
-
     # make a `_PassthroughScorer` with `check_scoring`:
-    scorer = check_scoring(meta_est, None)
-    scorer.set_score_request(sample_weight=True)
+    scorer = check_scoring(LogisticRegression(), None)
+    scorer.set_score_request(sample_weight='my_weights')
 
-    assert str(scorer.get_metadata_routing()) == str(scorer._metadata_request)
+    assert scorer.get_metadata_routing().score.requests['sample_weight'] == 'my_weights'
```

Reply:
Thanks @adrinjalali, but I find this very confusing. What is the behaviour we want `_PassthroughScorer` to have and WHY?

Consider this test, where I have tried to unify this new test with the one right above:

```python
    """Test that _PassthroughScorer.set_score_request adds the correct metadata request
    on itself."""
    # make a `_PassthroughScorer` with `check_scoring`:
    scorer = check_scoring(LogisticRegression().set_score_request(sample_weight="alias"), None)
    scorer.set_score_request(sample_weight='my_weights')
    assert scorer.get_metadata_routing().score.requests['sample_weight'] == 'my_weights'
```

When we would **not** do the `scorer.set_score_request(sample_weight='my_weights')`, then we would want to assert that sample_weight == "alias". At least this is what is done in the test above. Obviously, `_PassthroughScorer`s functionality has changed and maybe the test above is not valid anymore. But what are we trying to archive here?

First, we had detected that `RecursionError` on some (Stacking* is now also affected) metaestimators, then we fixed this by making `_PassthroughScorer` less passing-through, but here we are testing on normal consumers (LogisticRegression) and anything that has a score method.

When an estimator uses the default scoring, `_PassthroughScorer` should actually not change routed metadata, but if we do a `set_score_request` on it, it definitely changes routed metadata and we like that and want to test if it really does??

Reply:
You raise a good point there about carrying the request from the original estimator. So the more complete test would be:

```py
    est = LogisticRegression().set_score_request(sample_weight="alias")
    scorer = check_scoring(est, None)
    assert scorer.get_metadata_routing().score.requests['sample_weight'] == 'alias'
    scorer.set_score_request(sample_weight='my_weights')
    assert scorer.get_metadata_routing().score.requests['sample_weight'] == 'my_weights'
    # making sure changing the passthrough object doesn't affect the estimator.
    assert est.get_metadata_routing().score.requests['sample_weight'] == 'alias'
```

This is about having a correct public API, it's not about just making codecov happy.

Note that we're also making sure we have a non-regression test for the original issue. This is another test to make sure the added functionality behaves correctly.

Reply:
Okay, I think I now see what we want. Let me sum up:

We want a default scorer (`_PassthroughScorer`), that adopts the routing set on the estimator it scores for, but that doesn't force the estimator to change its own routing.
Since the default scorer is now newly inheriting from `_MetadataRequester` it now newly acts as a real consumer (while before this was only as if) and we can set its own routings. This is part of the public API, because `check_scoring` is.

Unfortunately however, `_PassthroughScorer` does change its estimators routing. The last assert fails.

I will try to find out why.

Reply:
Two things here are catching my attention and maybe one of them causes the test to fail:

1) `self` from within `_PassthroughScorer` is `<class 'sklearn.linear_model._logistic.LogisticRegression'>.score`.

2) When `requests.score = estimator._metadata_request.score` in is ran in `_PassthroughScorer.__init__`, the `MethodMetadataRequest` that is added to here is still belonging to the estimator (owner attribute).